### PR TITLE
OCLOMRS-238: Improve current implementation of Q-AND-A mapping

### PR DESCRIPTION
**Pull request by AGMETEOR**
# JIRA TICKET NAME:
[Improve current implementation of Q-AND-A mapping](https://issues.openmrs.org/browse/OCLOMRS-238)

# Summary:
The current implementation of the Q-AND-A mapping relationship creates a relationship based on "Narrower than" or "Same as".  This is not the type of mapping relationship for Q and A mappings, the relationship description for this type of mapping should be "Q-AND-A".
This ticket ideally might require taking out the old implementation.

### CODE REVIEW

**Comment by ODORA0** [comment]
@AGMETEOR Did you follow steps from here https://wiki.openmrs.org/display/docs/Pull+Request+Tips?

**Comment by AGMETEOR** [comment]
@ODORA0 The WIP at the beginning of the ticket was simply an indication that's a Work In Progress so that team members don't review it yet.

**Comment by dkayiwa** [comment]
These changes are too many. What has all this got to do with "the relationship description for this type of mapping should be "Q-AND-A"?

**Comment by AGMETEOR** [comment]
@dkayiwa the changes required a complete re-implementation as the initial feature to create Q-AND-A mappings was not what the POs wanted.

**Comment by dkayiwa** [comment]
Would you separate removal of the old implementation from adding the new one in separate pull requests?

**Comment by AGMETEOR** [comment]
@dkayiwa for this particular one no, because, there was a lot of code in the old implementation that I simply just had to change and in other cases replace in order to get the desired implementation.

File: src/components/dictionaryConcepts/components/AnswerRow.jsx [hunk 1]
```diff
@@ -1,59 +1,123 @@
...
+              />
+            ) : (
+              <select
+                name="map_scope"
+                className="form-control"
+                onChange={event => this.handleChangeInSource(event)}
+                required
+              >
+                <option value={localStorage.getItem('dictionaryId')}>{currentDictionaryName}</option>
+                <option value="CIEL">CIEL</option>
```
Comment (dkayiwa) [inline src/components/dictionaryConcepts/components/AnswerRow.jsx]: Don't we have a constant for the above?

File: src/components/dictionaryConcepts/containers/EditConcept.jsx [hunk 2]
```diff
@@ -338,6 +346,42 @@ export class EditConcept extends Component {
     });
   }
 
+  handleAsyncSelectChange = (answers, uniqueKey) => {
+    const { addSelectedAnswers, selectedAnswers } = this.props;
+    addSelectedAnswers(answers, uniqueKey);
+
+    const answersToAdd = selectedAnswers
+      .filter(ans => ans.map_type === 'Q-AND-A')
```
Comment (dkayiwa) [inline src/components/dictionaryConcepts/containers/EditConcept.jsx]: Should we hard code this?

File: src/components/dictionaryConcepts/components/AnswerRow.jsx (hunk 1 above)
Comment (AGMETEOR) [reply src/components/dictionaryConcepts/components/AnswerRow.jsx]: @dkayiwa  let me fix this.

File: src/components/dictionaryConcepts/containers/EditConcept.jsx (hunk 2 above)
Comment (AGMETEOR) [reply src/components/dictionaryConcepts/containers/EditConcept.jsx]: @dkayiwa let me create a variable for this.

File: src/redux/reducers/ConceptReducers.js [hunk 3]
```diff
@@ -24,15 +25,19 @@ import {
...
+  PRE_POPULATE_ANSWERS,
+  UNPOPULATE_PRE_POPULATED_ANSWERS,
+  ADD_NEW_ANSWER_ROW,
 } from '../actions/types';
 import {
   filterSources, filterClass, filterList, normalizeList, filterNames,
   filterDescriptions,
 } from './util';
 
+// const frontEndUniqueKey = uuid();
```
Comment (dkayiwa) [inline src/redux/reducers/ConceptReducers.js]: What is the above used for?

File: src/redux/reducers/ConceptReducers.js [hunk 4]
```diff
@@ -1,3 +1,4 @@
+// import uuid from 'uuid';
```
Comment (dkayiwa) [inline src/redux/reducers/ConceptReducers.js]: What is the above used for?

File: src/redux/reducers/ConceptReducers.js (hunk 3 above)
Comment (AGMETEOR) [reply src/redux/reducers/ConceptReducers.js]: @dkayiwa I have taken out the commented out code.

File: src/redux/reducers/ConceptReducers.js (hunk 4 above)
Comment (AGMETEOR) [reply src/redux/reducers/ConceptReducers.js]: @dkayiwa I have taken out the commented out code.

File: src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js [hunk 5]
```diff
@@ -94,6 +99,26 @@ describe('Test suite for dictionary concept actions', () => {
...
   });
 
+  it('should query possible answer concepts', async () => {
+    const expectedAnswers = [{ id: 'Anwer 1' }, { id: 'Anwer 2' }];
+
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith({ status: 200, response: expectedAnswers });
+    });
+    await queryAnswers('source', 'query');
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js]: Where are you asserting or expecting?

File: src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js [hunk 6]
```diff
@@ -94,6 +99,26 @@ describe('Test suite for dictionary concept actions', () => {
...
+  });
+
+  it('should query possible answer concepts from CIEL', async () => {
+    const expectedAnswers = [{ id: 'CIEL 1' }, { id: 'CIEL 2' }];
+
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith({ status: 200, response: expectedAnswers });
+    });
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js]: Where are you asserting or expecting?

Comment (AGMETEOR) [reply src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js]: @dkayiwa this was an error. I have fixed it now.

File: src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js (hunk 5 above)
Comment (AGMETEOR) [reply src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js]: @dkayiwa this was an error. I have fixed it now.

**Pull request merged**